Rewrite accounts and registration

kuba opened the pull request:
Save accounts to:

    /etc/letsencrypt/accounts/www.letsencrypt-dmeo.org/acme/new-reg/ \
    my.local.hostname@2015-07-04T14:04:10Z/ \
    {regr.json,meta.json,private_key.json}

Account now represents a combination of private key, Registration Resource and client account metadata. `Account.id` based on the account metadata (creation host and datetime). UI interface (`cli._determine_account`) based on the `id`, and not on email as previously.

Add `AccountStorage` interface and `AccountFileStorage`, `AccountMemoryStorage` implementations (latter, in-memory, useful for testing).

Create Account only after Registration Resource is received (`register()` returns `Account`).

Allow `client.Client(..., acme=acme, ...)`: API client might reuse `acme.client.Client` as returned by `register()`.

Move `report_new_account` to `letsencrypt.account`, `client.Client.register` into `client.register`.

Use `Registration.from_data` acme API.

`achallenges.AChallenge.key` is now the `acme.jose.JWK`, not `le_util.Key`. Plugins have to export PEM/DER as necessary (c.f. `letsencrypt.plugins.common.Dvsni.get_key_path`)

Add `--agree-tos`, save `--agree-eula` to `args.eula`. Prompt for EULA as soon as client is launched, add prompt for TOS.

Remove unnecessary `letsencrypt.network`. Remove, now irrelevant, `IConfig.account_keys_dir`.

Based on the draft from https://github.com/letsencrypt/letsencrypt/pull/362#issuecomment-97946817.

Also, expose `le_util.safe_open`.

Sorry about the size, @jdkasten!

jdkasten commented on letsencrypt/account.py:
```diff
@@ -22,212 +23,179 @@
-        else:
-            return None
+        """
+        return self.slug
```
self.slug()?

jdkasten replied:
Sorry missed that it was a property.  I do think that the hash of the account key is a bit easier to understand.

jdkasten commented on letsencrypt/cli.py:
```diff
@@ -468,8 +495,14 @@ def create_parser(plugins, args):
     helpful.add(
-        "automation", "--agree-eula", "-e", dest="tos", action="store_true",
+        "automation", "--agree-eula", dest="eula", action="store_true",
+        help="Agree to the Let's Encrypt Developer Preview EULA")
```
This will just be temporary.

jdkasten commented on letsencrypt/cli.py:
```diff
@@ -468,8 +495,14 @@ def create_parser(plugins, args):
         help="Agree to the Let's Encrypt Subscriber Agreement")
+    helpful.add(
+        "automation", "--account", metavar="ACCOUNT_ID",
+        help="Account ID to use")
```
Follow-up feature request  --list-accounts

kuba replied:
You meant `letsencrypt accounts` (similarily to `letsencrypt plugins`) :)

jdkasten commented on letsencrypt/cli.py:
```diff
@@ -724,6 +757,13 @@ def main(cli_args=sys.argv[1:]):
     zope.component.provideUtility(report)
     atexit.register(report.atexit_print_messages)
 
+    # TODO: remove developer EULA prompt for the launch
```
:+1:

jdkasten commented on letsencrypt/account.py:
```diff
@@ -1,233 +1,203 @@
+    reporter.add_message(recovery_msg, reporter.HIGH_PRIORITY, True)
+
+
+class AccountMemoryStorage(interfaces.AccountStorage):
```
I like the idea of the interface...though I don't see much use for the MemoryStorage implementation.  Unittests have to be written for the filebased system anyway and integration tests are already creating temporary directories.

kuba replied:
It's already being used in the unittests...

kuba commented:
Okay, changed `Account.id` to to be MD5 fingerprint, and included part of it in the `Account.slug`. Also, fixed `find_all` in case of `id` mismatch. For the record, account is now saved to:
```
/etc/letsencrypt/accounts/www.letsencrypt-demo.org/acme/new-reg/ \
24f0419708409ac6b3ec832582b700b5/ \
{regr.json,meta.json,private_key.json}
```

PTAL

jdkasten commented on letsencrypt/client.py:
```diff
@@ -30,48 +31,107 @@
+        `.Registration.terms_of_service` is accepted by the client, and
+        ``False`` otherwise. ``tos_cb`` will be called only if the
+        client acction is necessary, i.e. when ``terms_of_service is not
+        None``. This argument is optional, if not supplied it will
```
I think automatic acceptance will get us into trouble.  Hopefully this is intended for testing.

jdkasten commented on letsencrypt/client.py:
```diff
@@ -30,48 +31,107 @@
-
-        self.config = config
+        # Initialize ACME if account is provided
+        if acme is None and self.account is not None:
```
This seems a bit confusing.  The whole acme is created but only if you have created an account.  I do see the utility in creating one on the fly though....

kuba replied:
You need an Account Key to create acme. In fact we should have `assert account is not None` here, but this requires a whole `client.py` rewrite that I'll be working on next :)

jdkasten commented on letsencrypt/plugins/common.py:
```diff
@@ -157,19 +159,33 @@ def get_cert_file(self, achall):
+            encryption_algorithm=serialization.NoEncryption())
+        with le_util.safe_open(key_path, 'wb', chmod=0o400) as key_file:
+            key_file.write(key_pem)
+        self.configurator.reverter.register_file_creation(True, key_path)
```
This should happen before the file is written

jdkasten replied:
In case of a crash... it doesn't hurt if the reference has been created in the reverter... but it does hurt if the file was created and there is no reference to clean it up.

jdkasten replied:
On another level.... we probably shouldn't be writing out private keys that we don't need to. (The whole secure delete problem).  It seems like we should be able to avoid the write in this case.  It might require more thought though.

kuba replied:
This will have to be updated to recent spec changes anyway (#597). Account Key will no longer be used here, we will have to generate fresh new key.

jdkasten replied:
Agreed. Thanks for moving the registration before.

kuba commented:
Travis again hit by https://github.com/letsencrypt/boulder/issues/420

coolaj86 commented:
Could SHA1 or SHA256 instead of MD5? The WebCrypto standard doesn't support MD5 and it's widely being deprecated. I realized that it's not necessarily an issue of security, but when it gets implemented in other JavaScript environments (say a browser that acts as a file system), which will eventually happen, it'll be much easier to use a SHA-1.

coolaj86 commented:
And on that note, it looks like I'm a day late to the party, someone is already doing it:
https://groups.google.com/a/letsencrypt.org/forum/#!topic/ca-dev/OVNjIdM8fcc
